### ISSUE TITLE
[Bug]: different logprobs for qwn2-vl when running on transformers and on vllm

### ISSUE THREAD

**Issue by omercohen-glossai** [vllm-project/vllm#12699]
### Your current environment

Im runnig a finetuned qwen2-vl with transformers and with vllm and getting significantly different logprobs.
vllm is running on runpod vllm-worker and the environment data is in the attached file:
[pod_env.txt](https://github.com/user-attachments/files/18644453/pod_env.txt)
tranformers env data:
Package                           Version
--------------------------------- ---------------
accelerate                        0.34.0
annotated-types                   0.7.0
anyio                             4.2.0
argon2-cffi                       23.1.0
argon2-cffi-bindings              21.2.0
arrow                             1.3.0
asttokens                         2.4.1
async-lru                         2.0.4
attrs                             23.2.0
av                                14.1.0
Babel                             2.14.0
beautifulsoup4                    4.12.3
bleach                            6.1.0
blinker                           1.4
certifi                           2024.2.2
cffi                              1.16.0
charset-normalizer                3.3.2
click                             8.1.8
comm                              0.2.1
cryptography                      3.4.8
dbus-python                       1.2.18
debugpy                           1.8.0
decorator                         5.1.1
defusedxml                        0.7.1
distro                            1.7.0
docker-pycreds                    0.4.0
einops                            0.8.0
entrypoints                       0.4
exceptiongroup                    1.2.0
executing                         2.0.1
fastjsonschema                    2.19.1
filelock                          3.13.1
flash-attn                        2.7.2.post1
fqdn                              1.5.1
fsspec                            2024.2.0
gitdb                             4.0.12
GitPython                         3.1.44
h11                               0.14.0
httpcore                          1.0.2
httplib2                          0.20.2
httpx                             0.26.0
huggingface-hub                   0.27.1
idna                              3.6
importlib-metadata                4.6.4
ipykernel                         6.29.0
ipython                           8.21.0
ipython-genutils                  0.2.0
ipywidgets                        8.1.1
isoduration                       20.11.0
jedi                              0.19.1
jeepney                           0.7.1
Jinja2                            3.1.3
joblib                            1.4.2
json5                             0.9.14
jsonpointer                       2.4
jsonschema                        4.21.1
jsonschema-specifications         2023.12.1
jupyter-archive                   3.4.0
jupyter_client                    7.4.9
jupyter_contrib_core              0.4.2
jupyter_contrib_nbextensions      0.7.0
jupyter_core                      5.7.1
jupyter-events                    0.9.0
jupyter-highlight-selected-word   0.2.0
jupyter-lsp                       2.2.2
jupyter-nbextensions-configurator 0.6.3
jupyter_server                    2.12.5
jupyter_server_terminals          0.5.2
jupyterlab                        4.1.0
jupyterlab_pygments               0.3.0
jupyterlab_server                 2.25.2
jupyterlab-widgets                3.0.9
keyring                           23.5.0
launchpadlib                      1.10.16
lazr.restfulclient                0.14.4
lazr.uri                          1.0.6
lxml                              5.1.0
MarkupSafe                        2.1.5
matplotlib-inline                 0.1.6
mistune                           3.0.2
more-itertools                    8.10.0
mpmath                            1.3.0
nbclassic                         1.0.0
nbclient                          0.9.0
nbconvert                         7.14.2
nbformat                          5.9.2
nest-asyncio                      1.6.0
networkx                          3.2.1
notebook                          6.5.5
notebook_shim                     0.2.3
numpy                             1.26.3
nvidia-cublas-cu12                12.4.5.8
nvidia-cuda-cupti-cu12            12.4.127
nvidia-cuda-nvrtc-cu12            12.4.127
nvidia-cuda-runtime-cu12          12.4.127
nvidia-cudnn-cu12                 9.1.0.70
nvidia-cufft-cu12                 11.2.1.3
nvidia-curand-cu12                10.3.5.147
nvidia-cusolver-cu12              11.6.1.9
nvidia-cusparse-cu12              12.3.1.170
nvidia-nccl-cu12                  2.21.5
nvidia-nvjitlink-cu12             12.4.127
nvidia-nvtx-cu12                  12.4.127
oauthlib                          3.2.0
overrides                         7.7.0
packaging                         23.2
pandocfilters                     1.5.1
parso                             0.8.3
pexpect                           4.9.0
pillow                            10.2.0
pip                               24.0
platformdirs                      4.2.0
prometheus-client                 0.19.0
prompt-toolkit                    3.0.43
protobuf                          5.29.3
psutil                            5.9.8
ptyprocess                        0.7.0
pure-eval                         0.2.2
pycparser                         2.21
pydantic                          2.10.6
pydantic_core                     2.27.2
Pygments                          2.17.2
PyGObject                         3.42.1
PyJWT                             2.3.0
pyparsing                         2.4.7
python-apt                        2.4.0+ubuntu2
python-dateutil                   2.8.2
python-json-logger                2.0.7
PyYAML                            6.0.1
pyzmq                             24.0.1
qwen-vl-utils                     0.0.10
referencing                       0.33.0
regex                             2024.11.6
requests                          2.31.0
rfc3339-validator                 0.1.4
rfc3986-validator                 0.1.1
rpds-py                           0.17.1
safetensors                       0.5.2
scikit-learn                      1.6.1
scipy                             1.15.1
SecretStorage                     3.3.1
Send2Trash                        1.8.2
sentry-sdk                        2.20.0
setproctitle                      1.3.4
setuptools                        69.0.3
six                               1.16.0
smmap                             5.0.2
sniffio                           1.3.0
soupsieve                         2.5
stack-data                        0.6.3
sympy                             1.13.1
terminado                         0.18.0
threadpoolctl                     3.5.0
tinycss2                          1.2.1
tokenizers                        0.21.0
tomli                             2.0.1
torch                             2.5.1
torchaudio                        2.2.0
torchvision                       0.20.1
tornado                           6.4
tqdm                              4.67.1
traitlets                         5.14.1
transformers                      4.48.1
triton                            3.1.0
types-python-dateutil             2.8.19.20240106
typing_extensions                 4.12.2
uri-template                      1.3.0
urllib3                           2.2.0
wadllib                           1.3.6
wandb                             0.19.5
wcwidth                           0.2.13
webcolors                         1.13
webencodings                      0.5.1
websocket-client                  1.7.0
wheel                             0.42.0
widgetsnbextension                4.0.9
zipp                              1.0.0

### 🐛 Describe the bug

prompt:
[prompt.txt](https://github.com/user-attachments/files/18644904/prompt.txt)

vllm logprobs:
token 0:
	token=True, logprob=-0.3891778588294983
	token=False, logprob=-1.1391777992248535
	token=Yes, logprob=-7.0141777992248535
	token=true, logprob=-8.139178276062012
	token=false, logprob=-8.514178276062012
token 1:
	token=, logprob=-0.0007429938414134085
	token=., logprob=-7.8757429122924805
	token= False, logprob=-9.00074291229248
	token=,, logprob=-9.87574291229248
	token= True, logprob=-9.93824291229248

transformers logprobs:
  Logprob: -0.6362
  token id: 4049
  Top Logprobs: {'False': -0.6361945271492004, 'True': -0.7611945271492004, 'Yes': -6.886194705963135, 'false': -7.636194705963135, 'true': -7.886194705963135}

  Logprob: -0.0007
  token id: 151645
  Top Logprobs: {'<|im_end|>': -0.0007362039177678525, '.': -7.750736236572266, '\n': -9.000736236572266, '\n\n': -9.375736236572266, ',': -10.750736236572266}

When I insert prompts without images im getting similar logprobs so I think there is something wrong with the vision encoding.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
Can you show the code you used for HF? Also, does this issue happen in offline inference as well?

**Comment by omercohen-glossai** [comment]
1. This is also happening in offline inference - exactly same logprobs.
2. the hf code:
```
def print_vision_embeddings(given_model, given_processor,input_text):
    tokenizer = given_processor.tokenizer
    text = given_processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    # Load the tokenizer and model
    image_inputs, video_inputs = process_vision_info(messages)
    # Prepare the input for the model
    inputs = given_processor(
        text=[text],
        images=image_inputs,
        videos=video_inputs,
        padding=True,
        return_tensors="pt",
    )
    _, _ = inputs.pop('input_ids'), inputs.pop('attention_mask')
    # with torch.no_grad():
    #     given_model.

def print_top_k_log_probs(given_model, given_processor, input_text, k=5):
    """
    Prints the top k log probabilities for all output tokens of a Transformers model.

    Args:
        model_name (str): The name of the pretrained model to use.
        input_text (str): The input text to feed into the model.
        k (int): The number of top log probabilities to print per token.
    """
    tokenizer = given_processor.tokenizer
    text = given_processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    # Load the tokenizer and model
    image_inputs, video_inputs = process_vision_info(messages)
    # Prepare the input for the model
    inputs = given_processor(
        text=[text],
        images=image_inputs,
        videos=video_inputs,
        padding=True,
        return_tensors="pt",
    )
    # print(len(inputs['input_ids'][0]))
    # for e in inputs['input_ids'][0]:
    #     print(e.item())
    #print(inputs.input_ids.shape)
    # Ensure inputs are moved to the right device
    inputs = inputs.to("cuda")
    with torch.no_grad():  # Ensure no gradients are computed
        outputs = given_model.generate(**inputs,
                                       output_logits=True,
                                       return_dict_in_generate=True,
                                       output_scores=True,
                                       do_sample=False,
                                       use_cache=False,
                                       repetition_penalty= None,
                                       top_p=None,
                                       top_k=None,
                                       temperature= None,
                                       output_hidden_states=True
                                      )
    # print(outputs)
    # Get scores (logits -> log probs)
    scores = outputs.logits  # List of tensors for each step
    # for s in scores:
    #     print(s[s > 0])
    generated_token_ids = outputs.sequences[0][inputs.input_ids.size(1):]  # New tokens only

    results = []

    for i, token_id in enumerate(generated_token_ids):
        # Convert logits to log probabilities
        logits = scores[i][0]  # Take first batch
        log_probs = torch.log_softmax(logits, dim=-1)
        
        # Get log prob of the generated token - now correctly indexed
        token_logprob = log_probs[token_id].item()
        
        # Get top-k tokens and their log probs
        top_k_log_probs, top_k_token_ids = torch.topk(log_probs, k=k)
        
        # Create dictionary of top tokens and their log probs
        top_logprobs = {
            tokenizer.decode([tid.item()]): prob.item()
            for tid, prob in zip(top_k_token_ids, top_k_log_probs)
        }
        
        results.append({
            "token": tokenizer.decode([token_id]),
            "token_id": token_id,
            "logprob": token_logprob,
            "top_logprobs": top_logprobs
        })
    return results
```

**Comment by DarkLight1337** [comment]
It looks like you applied `process_vision_info` on HF side. Did you do the same for vLLM?

**Comment by omercohen-glossai** [comment]
I tried again, using `process_vision_info`  with vllm, and i got the same results.:
`tokenizer = processor.tokenizer
text = processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
image_inputs, video_inputs = process_vision_info(messages)
sampling_params = SamplingParams(logprobs=5, temperature=0, top_k=1, repetition_penalty=1.0, top_p=0.001)
res = vllm_model.generate(
    {
        "prompt":text,
        "multi_modal_data": {
                "image": image_inputs
        },
    },
    sampling_params=sampling_params
)`

**Comment by DarkLight1337** [comment]
Are the generated outputs different? Or just the logprobs?

**Comment by omercohen-glossai** [comment]
any idea?

**Comment by ywang96** [comment]
I'm not sure about this actually, because the qwen2-vl PR came directly from Qwen team, so maybe the team can comment on the implementation itself cc @fyabc 

@Isotr0py Do you know if this has anything to do with changes you made to vision encoder (if any?)

**Comment by Isotr0py** [comment]
I remember there has been a discussion about the degraded accuracy: https://github.com/vllm-project/vllm/issues/8408#issuecomment-2350938528. Would it be related to this?

**Comment by omercohen-glossai** [comment]
If I understand correctly, the implementation differences laid out in the mentioned comment results in very small numerical differences. In my case differences  between transformers and vllm are significant and im having different classification.

**Comment by ywang96** [comment]
@omercohen-glossai Do you also observe large logprobs if you only give the text-only input? (trying to rule out if this is an issue of the LM or ViT)

**Comment by omercohen-glossai** [comment]
After double checking, i notice differences in only-text input. Is there any script that compares transformers and vllm outputs that I can be 100% sure that the models are identically configured?

**Comment by ywang96** [comment]
> i notice differences in only-text input

Hmm so this means there's either something wrong with our Qwen2 LM implementation or the configuration of how models are set up (I'm more leaning towards the latter becasue otherwise we would have received a lot more bug reports since Qwen2 is pretty popular)

I don't have a script but out model tests do compare the output between HF & vLLM to a certain number of tokens, you can take a look at [test_models.py](https://github.com/vllm-project/vllm/blob/main/tests/models/decoder_only/vision_language/test_models.py), or run the test for Qwen2VL yourself with
```
pytest test_models.py -k qwen2_vl
```

**Comment by DarkLight1337** [comment]
Can you try running the model in full precision? https://github.com/vllm-project/vllm/issues/12343#issuecomment-2646767565 suggests that the discrepancies might stem from precision errors in the attention kernel.

**Comment by randomcodelookup** [comment]
> Hmm so this means there's either something wrong with our Qwen2 LM implementation or the configuration of how models are set up (I'm more leaning towards the latter becasue otherwise we would have received a lot more bug reports since Qwen2 is pretty popular)

I suspect this might be a widespread problem but not well reported, because people often use the model out of the box without inspecting the log probs.

> Can you try running the model in full precision? https://github.com/vllm-project/vllm/issues/12343#issuecomment-2646767565 suggests that the discrepancies might stem from precision errors in the attention kernel.

I also tested running the model in multiple dtypes `{auto,half,float16,bfloat16,float,float32}` and the log probs are still different. There's a big delta, but not as large as needed to make up for the original transformers difference. Don't think it's a precision issue

**Comment by omercohen-glossai** [comment]
I didnt manage to solve it and used another model eventually :)

**Comment by avinash31d** [comment]
I am facing the same issue with Phi4mm instruct model as well, 

the outputs are completely different.

Do you guys benchmark the PR for accuracy before merging?

**Comment by DarkLight1337** [comment]
How are you serving the model?

**Comment by avinash31d** [comment]
@DarkLight1337 I am using this example to run phi4mm
https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/vision_language_multi_image.py

**Comment by avinash31d** [comment]
@DarkLight1337 here is the HF transformers code for ph4mm
```

import requests
import torch
import os
import io
from PIL import Image
import soundfile as sf
from transformers import AutoModelForCausalLM, AutoProcessor, GenerationConfig
from urllib.request import urlopen
from peft import PeftModel


# Define model path
model_path = "./phi-4-multimodal-instruct"

# Load model and processor
processor = AutoProcessor.from_pretrained(model_path, trust_remote_code=True, use_fast=True)
model = AutoModelForCausalLM.from_pretrained(
            model_path,
                device_map="cuda",
                    torch_dtype=torch.bfloat16,
                        trust_remote_code=True,
                            # if you do not use Ampere or later GPUs, change attention to "eager"
                                _attn_implementation='flash_attention_2',
                                ).cuda()
generation_config = GenerationConfig.from_pretrained(model_path)

model = PeftModel.from_pretrained(model, "./phi-4-multimodal-instruct/vision-lora")

model = model.merge_and_unload()

system_prompt = '''
Put the system prompt here
'''

prompt= ""

# Define prompt structure
user_prompt = '<|user|>'
assistant_prompt = '<|assistant|>'
prompt_suffix = '<|end|>'

# Part 1: Image Processing
print("\n--- IMAGE PROCESSING ---")
image_url = 'image.png'
prompt = f'<|system|>{system_prompt}<|end|>{user_prompt}<|image_1|>{prompt}{prompt_suffix}{assistant_prompt}'
print(f'>>> Prompt\n{prompt}')

image = Image.open(image_url)
inputs = processor(text=prompt, images=image, return_tensors='pt').to('cuda:0')

import time
# Generate response

start_time = time.time()
generate_ids = model.generate(
        **inputs,
            max_new_tokens=100,
                generation_config=generation_config,
                num_logits_to_keep=0,
                do_sample=False
                )
generate_ids = generate_ids[:, inputs['input_ids'].shape[1]:]
response = processor.batch_decode(
        generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
        )[0]
total_run_time = time.time() - start_time
print(f'>>> Response\n{response}')
print(f"Total Time : {total_run_time:.2f}")
```

**Comment by DarkLight1337** [comment]
cc @Isotr0py I think you were looking into the incorrect results for multi image input?

**Comment by avinash31d** [comment]
@Isotr0py 
In vllm the value is 16 and in hf it is 36.
I corrected it to 36 but still output did not improve.

**Comment by Isotr0py** [comment]
Hmmm, I checked the vision embeddings in both vllm main branch and hf, and it's all close. 

Let me take a deeper investigation after #15477...
